Lilliefors' test for exponential distributions

jacobkimmel opened the pull request:
Updated the implementation of Lilliefors' test in `statsmodels.stats._lilliefors` to include Lilliefors' test for exponential distributions, in addition to the existing test for normality. 
The `TableDist` based implementation pattern in the existing module is followed.

The test for normality is retained as the default behavior when `statsmodels.stats._lilliefors.lilliefors()` is called to retain backwards compatibility.

`_lilliefors.kstest_normal` is renamed to `_lilliefors.kstest_lilliefors` to reflect the updated functionality.

This PR implements the following reference:

```
Lilliefors, H.W.
On the Kolmogorov-Smirnov test for the exponential distribution with mean unknown.
Journal of the American Statistical Association, Vol 64, No. 325. (1969), pp. 387–389.
```

Unit tests were added in `statsmodels/stats/tests/test_lilliefors.py`. 
There were no pre-existing tests prior to this PR.

josef-pkt commented:
@jacobkimmel  Thank you, 
The enhancement looks good
The only issue based on a quick read through is the renaming of the function name, kstest_normal.
First, for backwards compatibility we still need the old name that would need to be deprecated.
Second, I find the old name more informative, given that I never heard of Lilliefors before looking at the details, and general, but not strict avoidance of author names in function names. I'm not convinced yet of whether we should use the new name from this PR, use two different functions, kstest_normal and kstest_exponential, or find another generic name (kstest_fit ?)

we have #3673 which needs to be fixed for 0.9, and needs to be checked for the new expon version

(general aside: I have a PR that uses Stephens' tables or approximations for related gof tests)

jacobkimmel commented:
Understood, I realize namespace changes are a big deal!
I think creating two alias functions `kstest_normal` and `kstest_exponential` is a reasonable approach. `kstest_fit` may imply parameter estimation, which we don't do here. 
Just let me know which you'd like to implement & I'll add it to the PR.

I fixed #3673 in the recent commits. 
This was caused by invalid use of a p-value approximation function by the original author. The approximation is only valid where p < 0.1, but there were no bounds checks implemented. A test in `test_diagnostic.py` relied upon this faulty estimate. I've changed it to check for the correct p-value reported by the table. 

I've also added Lilliefors original tables to the normality function, which currently uses an alternative version from 

```
An Analytic Approximation to the Distribution of Lilliefors's Test Statistic for Normality
Gerard E. Dallal and Leland Wilkinson
The American Statistician 
```

These alternative tables add precision for p values < 0.01, but do not provide estimates for samples >>900. As such, samples >>900 currently cause an interpolation error for the normality test. Lilliefors tables allow for sampels >>900, but drop precision for p < 0.01. 

I've left the alternative tablles in place as before, but can easily switch to Lilliefors originals if desired.

josef-pkt commented on statsmodels/stats/_lilliefors.py:
```diff
@@ -87,55 +99,124 @@ def ksstat(x, cdf, alternative='two_sided', args=()):
 
     return lf
 
-lilliefors_table = get_lilliefors_table()
```
I think we should create two instances here, lllieforse_table_norm and ..._expon

jacobkimmel replied:
Will do! 👍

josef-pkt commented on statsmodels/stats/_lilliefors.py:
```diff
@@ -216,21 +308,37 @@ def kstest_normal(x, pvalmethod='approx'):
+        lilliefors_table = get_lilliefors_table(dist='norm')
+    elif dist == 'exp':
+        z = x
+        ex = stats.expon(scale=1./x.mean())
```
I think we can standardize z instead of using scale, as in the normal case

jacobkimmel replied:
Are you certain the tables are valid if we take that approach?

See the reference procedure in 

```
Lilliefors, H.W.
On the Kolmogorov-Smirnov test for the exponential distribution with mean unknown.
Journal of the American Statistical Association, Vol 64, No. 325. (1969), pp. 387–389.
```

which estimates the `lambda` parameter as the mean of `x`, rather than standardizing data as is typical for the normal case.

josef-pkt replied:
I meant standardize by scale z = x / x.mean()
Is your scale correct? I'm not sure whether it's x.mean or the reciprocal

jacobkimmel replied:
Ah! You're correct, my scale should be inverted! Nice catch.  
Lilliefors calls for `lambda = 1/X.mean()`

Then `scale = 1/lamda` for the usual parameterization. 

```
A common parameterization for expon is in terms of the rate parameter lambda, such that pdf = lambda * exp(-lambda * x). This parameterization corresponds to using scale = 1 / lambda.
```
https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.expon.html
 
I'll fix it.

That said, I don't believe scaling `x = x/x.mean()` by the mean is the equivalent to setting `lambda=x.mean()`.

Try the below as a demonstration. You'll get different distributions.

```
ex_dist = stats.expon(scale=1/0.5)
X_scaled = ex_dist.rvs(size=10000)
X_norm = X_scaled / X_scaled.mean()
X_estim = stats.expon(scale=1/(1./X_scaled.mean())).rvs(size=10000)
sns.distplot(X_norm)
sns.distplot(X_estim)
```

Here's the resulting plot.

![figure_1](https://user-images.githubusercontent.com/2465221/28698189-786b1ffa-72f6-11e7-8d5b-244ccb2c87b0.png)

Does that seem reasonable, or am I missing something here?


Thanks for the helpful feedback!

josef-pkt replied:
there is a difference in the scaled and unscaled "space", however these two should be equal
stats.expon(scale=1/(1./X_scaled.mean())).cdf(x) ==  stats.expon().cdf(x/X_scaled.mean()) 

the latter is 
stats.expon().cdf(z)   for z = x/X_scaled.mean() 
we I think we use

jacobkimmel replied:
Ah, understood! Changed to the normalization method.

josef-pkt commented on statsmodels/stats/tests/test_lilliefors.py:
```diff
@@ -0,0 +1,26 @@
+from statsmodels.stats._lilliefors import lilliefors
+import numpy as np
+
+class TestLilliefors(object):
```
The unit tests are not very strict. Are there some reference numbers from other packages available.

jacobkimmel replied:
I haven't been able to find another Python package that implements it, but there are `R` packages.

There are more stringent tests with non-randomized data in `test_diagnostic.py`. I'll add tests for the exponential function to those as well.

josef-pkt replied:
our unit tests usually verify the results against other packages mainly R or Stata if an equivalent function is available.
Those unit test should catch things like the wrong scale above.

jacobkimmel replied:
Understood, I'll get reference values from the R implementation and write new tests. Thanks for your guidance!

jacobkimmel commented:
@josef-pkt That last commit should include all the changes requested. 
The tests reference the `R` Lilliefors implementations in the `KScorrect` and `nortest` packages.

Did we decide on what to do about the namespace? Create `kstest_normal` and `kstest_exponential` aliases?

Otherwise -- I think this is wrapped up. 
Thanks for your help! This is my first PR to a reasonably sized project, so I appreciate your patience!

josef-pkt commented on statsmodels/stats/tests/test_lilliefors.py:
```diff
@@ -1,4 +1,19 @@
+
+Reference values from R packages:
+    `nortest` (lillie.test)
+    `KScorrect` (LcKS)
```
It's better to put this in a comment directly to the test
(in unit tests we use only `#` comments, not triple quotes)
Also adding the full R command or call is helpful for future reference if there are any options. 
(copy-paste into R is easier than figuring out what you might have used.)

jacobkimmel replied:
Fixed!

josef-pkt commented:
We haven't decided on function names yet. Otherwise looks good.

I will look at the details and API again, hopefully, tomorrow, so we can merge it.

josef-pkt commented:
This pull request introduces 3 alerts - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-de21a2c13a5ea667526650557306352e725c0ef6)

**new alerts:**

* 3 for Potentially uninitialized local variable

---

*Comment posted by [lgtm.com](https://lgtm.com)*

josef-pkt commented on statsmodels/stats/_lilliefors.py:
```diff
@@ -216,21 +313,36 @@ def kstest_normal(x, pvalmethod='approx'):
+        lilliefors_table = lilliefors_table_expon
+        pvalmethod = 'table'
+    else:
+        print('test distribution must be norm or exp.')
```
Thise should raise an exception, no print in code
ValueError or TypeError, I don't remember

jacobkimmel replied:
Fixed

josef-pkt commented on statsmodels/stats/tests/test_lilliefors.py:
```diff
@@ -0,0 +1,54 @@
+        d_ks_exp, p_exp = lilliefors(x_e, dist='exp')
+        # assert normal
+        assert_almost_equal(d_ks_norm, 0.15581, decimal=3)
+        assert_almost_equal(p_norm, 2.2e-16, decimal=3)
```
Why is the pvalue so low? Isn't it below all thresholds?

jacobkimmel replied:
The normality test uses the a p-value approximation on the low end (p < 0.1), so it's able to return p-values below those specified in the table. 
The exponential test doesn't have a comparable approximation function, so the lower bound is the same as in the table (p = 0.01). 

The cited reference is "Dallal Wilkinson 1986".

josef-pkt commented:
> I think creating two alias functions kstest_normal and kstest_exponential is a reasonable approach. kstest_fit may imply parameter estimation, which we don't do here.

We do estimate the parameters, mean and variance in the normal case and the rate parameter in the exponential case, that's why we need the special tables or approximation to the distribution, in contrast to scipy's one-sample ks-test which assumes all distribution parameters are fixed.

I don't know if we will add other distributions to this interface.
I suggest, calling the main function `kstest_fit` (or kstest_estimated) with the two aliases, kstest_normal and kstest_exponential (with keyword adjusted), plus the lilliefors alias. This adds two extra names, but leaves it open which ones we will eventually deprecate.
Maybe the docstring should mention that `kstest_fit` is the main function and the other ones are aliases or convenience functions for backwards compatibility and name recognition.
(something like kstest_fit sounds "future-proof" if we want to add more options.)

jacobkimmel commented:
> We do estimate the parameters, mean and variance in the normal case and the rate parameter in the exponential case, that's why we need the special tables or approximation to the distribution, in contrast to scipy's one-sample ks-test which assumes all distribution parameters are fixed.

Ah, I see your reasoning. My original comment was intended to convey that `*_fit` may imply that it returns parameters, but as you note, this isn't that different than any other model function that fits internally and outputs statistics.

I've changed the main function to `kstest_fit` and added the appropriate aliases `kstest_normal` and `kstest_exponential` (with the `dist='exp'` parameter specified using `functools.partial`) . 
I added a note to the top of the docstring to explain the namespace.

josef-pkt commented:
This pull request introduces 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-0da5ee46787b6bbd2051de88b021b0fa738ca3c3)

**new alerts:**

* 1 for Unreachable code

---

*Comment posted by [lgtm.com](https://lgtm.com)*

jacobkimmel commented:
Sorry, I'm not super familiar with appveyor or coveralls. It isn't apparent to me scanning the logs what's causing the decrease in test coverage or appveyor failures related to this PR. 

The tests that seem to fail in appveyor are outside any modules touched by this PR. Did something change in the testing suite between the initial submission and now?

josef-pkt commented:
Those are not problems related to this PR.

We just switched to using pytest instead of nosetests. 
An additional appveyor fail shows up now because it includes a test that it didn't include before (AFAIU), coverage numbers change because it now includes 0-coverage modules (modules in the sandbox that are not part of official and verified code and are not run by the unit tests at all)

jacobkimmel commented:
Got it - thanks!

josef-pkt commented:
Can you rebase on master? If not, I will do it.
I will try to go iver it one more time because I don't remember the details. Then I would like to merge it.

jacobkimmel commented:
That should be rebased!

josef-pkt commented:
This pull request introduces 5 alerts and fixes 8 - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-468eba094c34c61c46172e7c3bad7e525789744e)

**new alerts:**

* 2 for Module-level cyclic import
* 1 for __eq__ not overridden when adding attributes
* 1 for Unreachable code
* 1 for First argument to super() is not enclosing class

**fixed alerts:**

* 6 for Unreachable code
* 1 for Missing call to __init__ during object initialization
* 1 for __init__ method calls overridden method

---

*Comment posted by [lgtm.com](https://lgtm.com)*

josef-pkt commented:
I have a rebased version in #3936

josef-pkt commented:
merged rebased version in #3936
